sklearn/utils/_array_api.py

```diff
@@ -180,28 +187,31 @@ def _asarray_with_order(array, dtype=None, order=None, copy=None, xp=None):
     """
     if xp is None:
         xp, _ = get_namespace(array)
-    if xp.__name__ in {"numpy", "numpy.array_api"}:
+    if _is_numpy_namespace(xp):
         # Use NumPy API to support order
-        array = numpy.asarray(array, order=order, dtype=dtype)
-        return xp.asarray(array, copy=copy)
+        if copy is True:
+            array = numpy.array(array, order=order, dtype=dtype)
+        else:
+            array = numpy.asarray(array, order=order, dtype=dtype)
+        return xp.asarray(array)
     else:
         return xp.asarray(array, dtype=dtype, copy=copy)
 
 
 def _convert_to_numpy(array, xp):
```

Comment:
Why do I have to provide `xp` here? I was expecting it to be deduced from `array`.

Reply:
Maybe because we want to reuse the number of time we call `_get_namespace` when possible as it would introduce too much overhead? In particular as long as we need to rely on `_ArrayAPIWrapper` for some libraries.

Reply:
`_convert_to_numpy` is only used for testing, so I thought it was better to have the API be more explicit about the namespace.

I'm open to changing the signature in another PR, but I do not think it is a blocker for this PR.